src/main/java/com/cloudbees/jenkins/GitHubRepositoryNameContributor.java

```diff
@@ -36,41 +37,54 @@
      * then puts them into the collection.
      *
-     * @deprecated Use {@link #parseAssociatedNames(Job, Collection)}
+     * @deprecated Use {@link #parseAssociatedNames(Item, Collection)}
      */
     @Deprecated
     public void parseAssociatedNames(AbstractProject<?, ?> job, Collection<GitHubRepositoryName> result) {
-        parseAssociatedNames((Job) job, result);
+        parseAssociatedNames((Item) job, result);
     }
 
     /**
      * Looks at the definition of {@link Job} and list up the related github repositories,
      * then puts them into the collection.
+     * @deprecated Use {@link #parseAssociatedNames(Item, Collection)}
      */
+    @Deprecated
     public /*abstract*/ void parseAssociatedNames(Job<?, ?> job, Collection<GitHubRepositoryName> result) {
-        if (overriddenMethodHasDeprecatedSignature(job)) {
-            parseAssociatedNames((AbstractProject) job, result);
-        } else {
-            throw new AbstractMethodError("you must override the new overload of parseAssociatedNames");
-        }
+        parseAssociatedNames((Item) job, result);
     }
 
     /**
-     * To select backward compatible method with old extensions
-     * with overridden {@link #parseAssociatedNames(AbstractProject, Collection)}
-     *
-     * @param job - parameter to check for old class
-     *
-     * @return true if overridden deprecated method
+     * Looks at the definition of {@link Item} and list up the related github repositories,
+     * then puts them into the collection.
      */
-    private boolean overriddenMethodHasDeprecatedSignature(Job<?, ?> job) {
-        return Util.isOverridden(
+    @SuppressWarnings("deprecation")
+    public /*abstract*/ void parseAssociatedNames(Item item, Collection<GitHubRepositoryName> result) {
+        if (Util.isOverridden(
+                GitHubRepositoryNameContributor.class,
+                getClass(),
+                "parseAssociatedNames",
+                Job.class,
+                Collection.class
+        )) {
+            // if this impl is legacy, it cannot contribute to non-jobs, so not an error
+            if (item instanceof Job) {
+                parseAssociatedNames((Job<?, ?>) item, result);
+            }
+        } else  if (Util.isOverridden(
                 GitHubRepositoryNameContributor.class,
                 getClass(),
                 "parseAssociatedNames",
                 AbstractProject.class,
                 Collection.class
-        ) && job instanceof AbstractProject;
+        )) {
+            // if this impl is legacy, it cannot contribute to non-projects, so not an error
+            if (item instanceof AbstractProject) {
+                parseAssociatedNames((AbstractProject<?, ?>) item, result);
+            }
+        } else {
+            throw new AbstractMethodError("you must override the new overload of parseAssociatedNames");
```

Comment:
Also please check the previous implementation of `parseAssociatedNames(Job<?, ?> job, Collection<GitHubRepositoryName> result)` which threw the exact same exception. All I am doing here is moving the exception throwing from the previously most generic method to the new most generic method.

The exception will only be thrown if you have an implementation that does not override *any* `parseAssociatedNames` methods. If this had been designed as an SPI then the implementation methods would be protected and we could have shifted the abstract and used a guard.

To clarify for @KostyaSha this is not a bug just migration of existing pattern